platooning_tactical_plugin/include/platooning_tactical_plugin/platooning_tactical_plugin.h

```diff
@@ -1,104 +1,239 @@
-
-            // initialize this node
-            void initialize();
-
-            
-            double start_speed = 0;
-            double end_speed = 0;
-
-            // create uneven trajectory from waypoints
-            std::vector<cav_msgs::TrajectoryPlanPoint> create_uneven_trajectory_from_waypoints(std::vector<autoware_msgs::Waypoint> waypoints);
-
-            // get a sublist of waypoints marked by desired time span
-            std::vector<autoware_msgs::Waypoint> get_waypoints_in_time_boundary(std::vector<autoware_msgs::Waypoint> waypoints, double time_span);
-
-            // postprocess traj to add plugin names and shift time origin to the current ROS time
-            std::vector<cav_msgs::TrajectoryPlanPoint> post_process_traj_points(std::vector<cav_msgs::TrajectoryPlanPoint> trajectory);
-
-            // local copy of pose
-            boost::shared_ptr<geometry_msgs::PoseStamped const> pose_msg_;
-
-            /*!
-            * Destructor.
-            */
-            virtual ~PlatooningTacticalPlugin();
-
-        private:
-
-            // node handles
-            std::shared_ptr<ros::CARMANodeHandle> nh_, pnh_;
-
-            //! ROS subscribers.
-            ros::Subscriber waypoint_subscriber_;
-            ros::Subscriber pose_subscriber_;
-            ros::Subscriber twist_subscriber_;
-
-            // ros service servers
-            ros::ServiceServer trajectory_srv_;
-
-            // generated trajectory plan
-            cav_msgs::TrajectoryPlan trajectory_msg;
-            // Array of waypoints
-            std::vector<autoware_msgs::Waypoint> waypoints_list;
-
-            void waypoints_cb(const autoware_msgs::LaneConstPtr& msg);
-            void pose_cb(const geometry_msgs::PoseStampedConstPtr& msg);
-            void twist_cd(const geometry_msgs::TwistStampedConstPtr& msg);
-
-            bool plan_trajectory_cb(cav_srvs::PlanTrajectoryRequest &req, cav_srvs::PlanTrajectoryResponse &resp);
-
-            // Plugin discovery message
-            cav_msgs::Plugin plugin_discovery_msg_;
-            ros::Publisher platooning_tactical_plugin_discovery_pub_;
-
-            // ROS params
-            double trajectory_time_length_;
-            double trajectory_point_spacing_;
-
-            // current vehicle speed
-            double current_speed_;
-
-            // convert waypoints to a trajectory
-            std::vector<cav_msgs::TrajectoryPlanPoint> compose_trajectory_from_waypoints(std::vector<autoware_msgs::Waypoint> waypoints);
-
-    };
-}
\ No newline at end of file
+#include <carma_wm/WMListener.h>
+#include <functional>
+#include <platooning_tactical_plugin/smoothing/SplineI.h>
+
+#include "platooning_tactical_plugin_config.h"
+#include "third_party_library/spline.h"
+
+namespace platooning_tactical_plugin
+{
+using PublishPluginDiscoveryCB = std::function<void(const cav_msgs::Plugin&)>;
+
+/**
+ * \brief Convenience class for pairing 2d points with speeds
+ */ 
+struct PointSpeedPair
```

Comment:
Doesn't need to change, but wanted to comment that I think Mike has been using a cool technique for this kind of convenience struct: `using std::pair<TYPE_A, TYPE_B> as ...;`.